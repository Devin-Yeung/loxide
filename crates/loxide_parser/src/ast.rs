#[derive(Debug)]
pub struct Expr<'src> {
    pub kind: ExprKind<'src>,
}

#[derive(Debug)]
pub enum ExprKind<'src> {
    Literal(Literal<'src>),
    Unary(UnaryExpr<'src>),
    Binary(BinaryExpr<'src>),
    Grouped(Box<Expr<'src>>),
    Variable(Variable<'src>),
    Assign(AssignExpr<'src>),
}

#[derive(Debug)]
pub enum Literal<'src> {
    String(std::borrow::Cow<'src, str>),
    Number(f64),
    Boolean(bool),
    Nil,
}

#[derive(Debug)]
pub enum UnaryOperator {
    Minus,
    Bang,
}

#[derive(Debug)]
pub struct UnaryExpr<'src> {
    pub operator: UnaryOperator,
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub enum BinaryOperator {
    EqualEqual,
    BangEq,
    GreaterEqual,
    Greater,
    Less,
    LessEqual,
    Minus,
    Plus,
    Slash,
    Star,
}

#[derive(Debug)]
pub struct BinaryExpr<'src> {
    pub lhs: Box<Expr<'src>>,
    pub rhs: Box<Expr<'src>>,
    pub operator: BinaryOperator,
}

#[derive(Debug)]
pub struct GroupedExpr<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct AssignExpr<'src> {
    pub name: Variable<'src>,
    pub value: Box<Expr<'src>>,
}

impl<'src> AssignExpr<'src> {
    pub fn new(name: &'src str, value: Expr<'src>) -> Self {
        AssignExpr {
            name: Variable { name },
            value: Box::new(value),
        }
    }
}

#[derive(Debug)]
pub struct Variable<'src> {
    pub name: &'src str,
}

#[derive(Debug)]
pub struct ConditionStmt<'src> {
    pub condition: Expr<'src>,
    pub then_branch: Box<Stmt<'src>>,
    pub else_branch: Option<Box<Stmt<'src>>>,
}

#[derive(Debug)]
pub struct WhileStmt<'src> {
    pub condition: Expr<'src>,
    pub body: Box<Stmt<'src>>,
}

#[derive(Debug)]
#[non_exhaustive]
pub enum Stmt<'src> {
    Expression(Expr<'src>),
    PrintStmt(Expr<'src>),
    VarDeclaration(Variable<'src>, Option<Expr<'src>>),
    Block(Vec<Stmt<'src>>),
    Condition(ConditionStmt<'src>),
    While(WhileStmt<'src>),
}
