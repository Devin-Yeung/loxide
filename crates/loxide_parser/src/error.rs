use thiserror::Error;

#[derive(Error, Debug, Eq, PartialEq, Copy, Clone)]
pub enum SyntaxError {
    #[error("Unexpected character at line {0}")]
    UnexpectedChar(usize),
    #[error("Unexpected EOF")]
    UnexpectedEOF,
    #[error("Expected")]
    Expect(&'static str),
    #[error("Invalid Conversion")]
    InvalidConversion,
    #[error("Invalid Number")]
    InvalidNumber,
    #[error("Invalid string literal")]
    InvalidStringLiteral,
    #[error("Invalid keyword")]
    InvalidKeywords,
}
