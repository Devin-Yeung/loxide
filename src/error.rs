use thiserror::Error;

#[derive(Error, Debug, Eq, PartialEq)]
pub enum SyntaxError {
    #[error("Unexpected character at line {0}")]
    UnexpectedChar(usize),
    #[error("Invalid Number")]
    InvalidNumber,
    #[error("Invalid string literal")]
    InvalidStringLiteral,
    #[error("Invalid keyword")]
    InvalidKeywords,
}
