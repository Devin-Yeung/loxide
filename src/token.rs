#[derive(Debug, PartialEq)]
pub enum Literal {
    Identifier,
    String(std::string::String),
    Number(f64),
}

#[derive(Debug, PartialEq)]
pub enum Keyword {
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
}

#[derive(Debug, PartialEq)]
pub enum TokenType {
    // Single-character tokens.
    /// Opening parentheses `(`
    LeftParen,
    /// Closing parentheses `)`
    RightParen,
    /// Opening curly brace `{`
    LeftBrace,
    /// Closing curly brace `}`
    RightBrace,
    /// Comma `,`
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Comment
    Comment,
    EOF,
    // Literals.
    Literal(Literal),
    // Keywords.
    Keyword(Keyword),
}

#[derive(Debug)]
pub struct Token<'src> {
    pub(crate) ty: TokenType,
    pub(crate) lexeme: &'src str,
    pub(crate) span: Span,
}

/// The span of a token, line number indexed from 1,
/// column number indexed from 0, right exclusive.
#[derive(Debug, Copy, Clone)]
pub struct Span {
    pub(crate) line: usize,
    pub(crate) start: usize,
    pub(crate) end: usize,
}

impl<'src> Token<'src> {
    pub fn new(ty: TokenType, lexeme: &'src str, span: Span) -> Token<'src> {
        Token { ty, lexeme, span }
    }
}
