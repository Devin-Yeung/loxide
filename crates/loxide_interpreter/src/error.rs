use crate::value::{Value, ValueKind};
use loxide_parser::token::Span;
use miette::Diagnostic;
use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[non_exhaustive]
pub enum RuntimeError {
    #[error("Invalid unary operand, expected: {1}")]
    InvalidUnaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Invalid binary operand, expected: {1}")]
    InvalidBinaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Incompatible operand, {lhs_ty} is incompatible with {rhs_ty}")]
    IncompatibleBinaryOperand {
        #[label("has type `{lhs_ty}` here")]
        lhs_span: Span,
        lhs_ty: ValueKind,
        #[label("has type `{rhs_ty}` here")]
        rhs_span: Span,
        rhs_ty: ValueKind,
    },
    #[error("Undefined Variable: {1}")]
    UndefinedVariable(#[label("variable `{1}` is not defined")] Span, String),
    #[error("Expected boolean in condition expression")]
    ExpectedBoolean(
        #[label("expect a boolean type here, but got type `{1}`")] Span,
        ValueKind,
    ),
    #[error("Return value")]
    ReturnValue(Value),
    #[error("Bad arity, expected: {expected}, found: {found}")]
    BadArity {
        #[label("function declare here with {expected} parameters")]
        callee_span: Span,
        #[label("function call here with {found} parameters")]
        caller_span: Span,
        expected: usize,
        found: usize,
    },
    #[error("Function call on a non-callable value")]
    CallOnNonCallable(
        #[label("expect a callable value here, but got type `{1}`")] Span,
        ValueKind,
    ),
    #[error("Return in top level")]
    ReturnInTopLevel(#[label("return statement is not allowed in top level")] Span),
    #[error("Too many parameters, allow at most 255, found {1}")]
    TooManyParameters(#[label("Can't have more than 255 parameters")] Span, usize),
    #[error("Too many arguments, allow at most 255, found {1}")]
    TooManyArguments(#[label("Can't have more than 255 arguments")] Span, usize),
}

#[derive(Debug)]
pub(crate) enum PrivateRuntimeError {
    /// Bad Arity Error without Span
    BadArity { expected: usize, found: usize },
    /// Transparent Error of `RuntimeError`
    Transparent(RuntimeError),
}
