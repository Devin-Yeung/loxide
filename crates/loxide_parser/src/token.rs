use crate::ast;
use crate::ast::{BinaryOperator, UnaryOperator};
use crate::error::SyntaxError;
use miette::SourceSpan;
use std::borrow::Cow;
use std::str::FromStr;

#[derive(Debug, PartialEq, Clone)]
pub enum Literal<'src> {
    String(Cow<'src, str>),
    Number(f64),
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Keyword {
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
}

impl FromStr for Keyword {
    type Err = crate::error::SyntaxError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "and" => Ok(Keyword::And),
            "class" => Ok(Keyword::Class),
            "else" => Ok(Keyword::Else),
            "false" => Ok(Keyword::False),
            "fun" => Ok(Keyword::Fun),
            "for" => Ok(Keyword::For),
            "if" => Ok(Keyword::If),
            "nil" => Ok(Keyword::Nil),
            "or" => Ok(Keyword::Or),
            "print" => Ok(Keyword::Print),
            "return" => Ok(Keyword::Return),
            "super" => Ok(Keyword::Super),
            "this" => Ok(Keyword::This),
            "true" => Ok(Keyword::True),
            "var" => Ok(Keyword::Var),
            "while" => Ok(Keyword::While),
            _ => Err(crate::error::SyntaxError::InvalidKeywords),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType<'src> {
    // Single-character tokens.
    /// Opening parentheses `(`
    LeftParen,
    /// Closing parentheses `)`
    RightParen,
    /// Opening curly brace `{`
    LeftBrace,
    /// Closing curly brace `}`
    RightBrace,
    /// Comma `,`
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Comment
    Comment,
    EOF,
    // Literals.
    Literal(Literal<'src>),
    // Keywords.
    Keyword(Keyword),
    Identifier,
}

impl<'src> TokenType<'src> {
    pub fn name(&self) -> &'static str {
        match self {
            TokenType::LeftParen => "(",
            TokenType::RightParen => ")",
            TokenType::LeftBrace => "{",
            TokenType::RightBrace => "}",
            TokenType::Comma => ",",
            TokenType::Dot => ".",
            TokenType::Minus => "-",
            TokenType::Plus => "+",
            TokenType::Semicolon => ";",
            TokenType::Slash => "/",
            TokenType::Star => "*",
            TokenType::Bang => "!",
            TokenType::BangEqual => "!=",
            TokenType::Equal => "=",
            TokenType::EqualEqual => "==",
            TokenType::Greater => ">",
            TokenType::GreaterEqual => ">=",
            TokenType::Less => "<",
            TokenType::LessEqual => "<=",
            TokenType::Comment => "comment",
            TokenType::EOF => "eof",
            TokenType::Identifier => "identifier",
            TokenType::Literal(_) => "literal",
            TokenType::Keyword(_) => "keyword",
        }
    }
}

#[derive(Debug)]
pub struct Token<'src> {
    pub(crate) ty: TokenType<'src>,
    pub(crate) lexeme: &'src str,
    pub(crate) span: Span,
}

/// The span of a token, indexed from 0, left inclusive, right exclusive.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl From<Span> for (usize, usize) {
    fn from(val: Span) -> Self {
        (val.start, val.end)
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Span { start, end }
    }
}

impl From<Span> for SourceSpan {
    fn from(val: Span) -> Self {
        SourceSpan::from(val.start..val.end)
    }
}

impl<'src> Token<'src> {
    pub fn new(ty: TokenType<'src>, lexeme: &'src str, span: Span) -> Token<'src> {
        Token { ty, lexeme, span }
    }
}

impl<'src> TryInto<BinaryOperator> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<BinaryOperator, Self::Error> {
        let op = match self.ty {
            TokenType::EqualEqual => BinaryOperator::EqualEqual,
            TokenType::BangEqual => BinaryOperator::BangEq,
            TokenType::GreaterEqual => BinaryOperator::GreaterEqual,
            TokenType::Greater => BinaryOperator::Greater,
            TokenType::Less => BinaryOperator::Less,
            TokenType::LessEqual => BinaryOperator::LessEqual,
            TokenType::Minus => BinaryOperator::Minus,
            TokenType::Plus => BinaryOperator::Plus,
            TokenType::Slash => BinaryOperator::Slash,
            TokenType::Star => BinaryOperator::Star,
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(op)
    }
}

impl<'src> TryInto<UnaryOperator> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<UnaryOperator, Self::Error> {
        let op = match self.ty {
            TokenType::Minus => UnaryOperator::Minus,
            TokenType::Bang => UnaryOperator::Bang,
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(op)
    }
}

impl<'src> TryInto<ast::Literal> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<ast::Literal, Self::Error> {
        let literal = match self.ty {
            TokenType::Literal(Literal::String(raw)) => ast::Literal::String(raw.to_string()),
            TokenType::Literal(Literal::Number(raw)) => ast::Literal::Number(raw),
            TokenType::Keyword(Keyword::Nil) => ast::Literal::Nil,
            TokenType::Keyword(Keyword::True) => ast::Literal::Boolean(true),
            TokenType::Keyword(Keyword::False) => ast::Literal::Boolean(false),
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(literal)
    }
}
