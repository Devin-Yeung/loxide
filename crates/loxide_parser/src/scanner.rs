use crate::error::SyntaxError;
use crate::prophetic::Prophet;
use crate::token::{Keyword, Literal, Span, Token, TokenType};
use std::borrow::Cow;
use std::str::Chars;

pub struct Scanner<'src> {
    src: &'src str,
    prophet: Prophet<Chars<'src>>,
    current: Span,
    /// Number of chars already have been consumed
    /// this field increased only when we yield a token
    consumed: usize,
}

impl<'src> Scanner<'src> {
    pub fn new(src: &'src str) -> Scanner<'src> {
        Scanner {
            src,
            prophet: Prophet::new(src.chars()),
            current: Span { start: 0, end: 0 },
            consumed: 0,
        }
    }

    pub fn consume_if(&mut self, expected: char) -> bool {
        if self.peek() == Some(expected) {
            self.advance();
            true
        } else {
            false
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.prophet.peek()
    }

    fn peek_next(&mut self) -> Option<char> {
        self.prophet.peek_next()
    }

    /// advance will take care of span and iter for you
    fn advance(&mut self) -> Option<char> {
        match self.prophet.peek() {
            Some(_) => {
                self.current.end += 1;
            }
            _ => { /* Do nothing */ }
        }
        self.prophet.next()
    }

    pub fn is_at_end(&self) -> bool {
        self.prophet.peek().is_none()
    }

    fn eat_whitespace(&mut self) {
        loop {
            match self.peek() {
                Some(' ') | Some('\n') | Some('\r') | Some('\t') => {
                    self.advance();
                    self.consumed += 1
                }
                _ => {
                    self.current.start = self.current.end;
                    break;
                }
            }
        }
    }

    pub fn scan_token(&mut self) -> Result<Token<'src>, SyntaxError> {
        self.eat_whitespace();
        let ty = match self.advance() {
            Some('(') => TokenType::LeftParen,
            Some(')') => TokenType::RightParen,
            Some('{') => TokenType::LeftBrace,
            Some('}') => TokenType::RightBrace,
            Some(',') => TokenType::Comma,
            Some('.') => TokenType::Dot,
            Some('-') => TokenType::Minus,
            Some('+') => TokenType::Plus,
            Some(';') => TokenType::Semicolon,
            Some('*') => TokenType::Star,
            Some('!') => match self.consume_if('=') {
                true => TokenType::BangEqual,
                false => TokenType::Bang,
            },
            Some('=') => match self.consume_if('=') {
                true => TokenType::EqualEqual,
                false => TokenType::Equal,
            },
            Some('<') => match self.consume_if('=') {
                true => TokenType::LessEqual,
                false => TokenType::Less,
            },
            Some('>') => match self.consume_if('=') {
                true => TokenType::GreaterEqual,
                false => TokenType::Greater,
            },
            Some('/') => match self.consume_if('/') {
                true => {
                    while self.peek() != Some('\n') && !self.is_at_end() {
                        self.advance();
                    }
                    TokenType::Comment
                }
                false => TokenType::Slash,
            },
            Some('\"') => self.string()?,
            Some(c) if c.is_ascii_digit() => self.number()?,
            Some(c) if c.is_ascii_alphabetic() || c == '_' => self.identifier(),
            None => TokenType::EOF,
            _ => return Err(SyntaxError::UnexpectedChar(self.current)),
        };
        Ok(self.yield_token(ty))
    }

    fn yield_token(&mut self, ty: TokenType<'src>) -> Token<'src> {
        let lexeme = self.lexeme();
        self.consumed += lexeme.len();
        Token {
            ty,
            lexeme,
            span: self.current,
        }
    }

    /// return the lexeme currently working on
    fn lexeme(&self) -> &'src str {
        let len = self.current.end - self.current.start;
        &self.src[self.consumed..self.consumed + len]
    }

    fn string(&mut self) -> Result<TokenType<'src>, SyntaxError> {
        loop {
            match self.advance() {
                None => return Err(SyntaxError::InvalidStringLiteral(self.current)),
                Some('\"') => {
                    // discard the left quote
                    // return owned string only when *string escape* happens
                    let lexeme = self.lexeme();
                    // trim the left and right quote
                    let raw_str = Cow::Borrowed(&lexeme[1..lexeme.len() - 1]);
                    return Ok(TokenType::Literal(Literal::String(raw_str)));
                }
                _ => { /* Continue */ }
            }
        }
    }

    fn number(&mut self) -> Result<TokenType<'src>, SyntaxError> {
        while let Some(c) = self.peek() {
            match c {
                '.' => {
                    if self.peek_next().map_or(false, |c| c.is_ascii_digit()) {
                        self.advance()
                    } else {
                        return Err(SyntaxError::InvalidNumber(Span {
                            start: self.current.start,
                            end: self.current.end + 1,
                        }));
                    }
                }
                c if c.is_ascii_digit() => self.advance(),
                _ => break,
            };
        }
        let num = self
            .lexeme()
            .parse::<f64>()
            .map_err(|_| SyntaxError::InvalidNumber(self.current))?;
        Ok(TokenType::Literal(Literal::Number(num)))
    }

    fn identifier(&mut self) -> TokenType<'src> {
        while let Some(c) = self.peek() {
            if c.is_ascii_alphanumeric() {
                self.advance();
            } else {
                break;
            }
        }
        let content =
            &self.src[self.consumed..self.consumed + self.current.end - self.current.start];
        content
            .parse::<Keyword>()
            .map(TokenType::Keyword)
            .unwrap_or(TokenType::Identifier)
    }
}

impl<'src> Iterator for Scanner<'src> {
    type Item = Result<Token<'src>, SyntaxError>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.scan_token() {
            Ok(x) if x.ty == TokenType::EOF => None,
            Ok(x) => Some(Ok(x)),
            Err(e) => Some(Err(e)),
        }
    }
}

impl<'src> From<&'src str> for Scanner<'src> {
    fn from(value: &'src str) -> Self {
        Scanner {
            src: value,
            prophet: Prophet::new(value.chars()),
            current: Span { start: 0, end: 0 },
            consumed: 0,
        }
    }
}

impl<'src> From<&'src String> for Scanner<'src> {
    fn from(value: &'src String) -> Self {
        Scanner {
            src: value,
            prophet: Prophet::new(value.chars()),
            current: Span { start: 0, end: 0 },
            consumed: 0,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::scanner::Scanner;
    use crate::utils::test_utils::{LINEBREAK, SPAN_FILTER};
    use loxide_diagnostic::reporter::{Reporter, Style};
    use loxide_testsuite::unittest;
    use miette::Report;
    use std::sync::Arc;

    macro_rules! tokens {
        ($src:expr) => {{
            let scanner = $crate::scanner::Scanner::from($src);
            scanner.collect::<::std::vec::Vec<_>>()
        }};
    }

    fn display_tokenize_error<S: AsRef<str>>(src: S) -> String {
        let mut reporter = Reporter::new(Style::NoColor);
        let source = Arc::new(src.as_ref().to_string());
        let errs = tokens!(src.as_ref())
            .into_iter()
            .filter(|x| x.is_err())
            .map(|x| {
                let report: Report = x.unwrap_err().into();
                report.with_source_code(source.clone())
            })
            .collect::<Vec<_>>();
        reporter.extend(errs);
        reporter.report_to_string()
    }

    unittest!(single_char, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.map(|x| x.unwrap()).collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(single_or_double_char, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.map(|x| x.unwrap()).collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(valid_number, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(valid_string, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(valid_idents_and_keywords, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(expressions, filters => vec![SPAN_FILTER], |src| {
        let scanner = Scanner::from(src);
        let tokens = scanner.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });

    unittest!(should_fail, |src| {
        let output = src
            .split(LINEBREAK)
            .map(display_tokenize_error)
            .collect::<Vec<_>>()
            .join("\n");

        insta::assert_snapshot!(output);
    });
}
