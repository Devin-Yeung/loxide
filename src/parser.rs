// expression     → equality ;
// equality       → comparison ( ( "!=" | "==" ) comparison )* ;
// comparison     → term ( ( ">" | ">=" | "<" | "<=" ) term )* ;
// term           → factor ( ( "-" | "+" ) factor )* ;
// factor         → unary ( ( "/" | "*" ) unary )* ;
// unary          → ( "!" | "-" ) unary
//                | primary ;
// primary        → NUMBER | STRING | "true" | "false" | "nil"
//                | "(" expression ")" ;

use crate::ast;
use crate::ast::ExprKind::Binary;
use crate::ast::Literal::{Boolean, Nil};
use crate::ast::{BinaryExpr, Expr, ExprKind, Stmt, UnaryExpr};
use crate::error::SyntaxError;
use crate::scanner::Scanner;
use crate::token::{Keyword, Token, TokenType};
use std::iter::Peekable;

pub struct Parser<'src> {
    tokens: Peekable<Scanner<'src>>,
}

impl<'src> Parser<'src> {
    fn new(src: &'src str) -> Parser {
        Parser {
            tokens: Scanner::new(src).peekable(),
        }
    }

    fn consume(&mut self, ty: TokenType) -> Result<(), SyntaxError> {
        return if self.peek_type()? == ty {
            self.advance()?;
            Ok(())
        } else {
            todo!()
        };
    }

    fn peek_type(&mut self) -> Result<TokenType<'src>, SyntaxError> {
        match self.tokens.peek() {
            None => Ok(TokenType::EOF),
            Some(&Ok(ref token)) => Ok(token.ty.clone()),
            Some(&Err(ref err)) => Err(err.clone()),
        }
    }

    fn advance(&mut self) -> Result<Token<'src>, SyntaxError> {
        self.tokens
            .next()
            .unwrap_or_else(|| Err(SyntaxError::UnexpectedEOF))
    }

    fn statement(&mut self) -> Result<Stmt<'src>, SyntaxError> {
        self.expression_stmt()
    }

    fn expression_stmt(&mut self) -> Result<Stmt<'src>, SyntaxError> {
        let expr = self.expression()?;
        self.consume(TokenType::Semicolon)?;
        Ok(Stmt::Expression(expr))
    }

    fn expression(&mut self) -> Result<Expr<'src>, SyntaxError> {
        self.equality()
    }

    fn equality(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let mut expr = self.comparison()?;
        loop {
            let token = match self.peek_type()? {
                TokenType::EqualEqual | TokenType::BangEqual => self.advance()?,
                _ => break,
            };
            let rhs = self.comparison()?;
            expr = Expr {
                kind: Binary(BinaryExpr {
                    lhs: Box::new(expr),
                    rhs: Box::new(rhs),
                    operator: token.try_into()?,
                }),
            }
        }
        Ok(expr)
    }

    fn comparison(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let mut expr = self.term()?;
        loop {
            let token = match self.peek_type()? {
                TokenType::Greater
                | TokenType::GreaterEqual
                | TokenType::Less
                | TokenType::LessEqual => self.advance()?,
                _ => break,
            };
            let rhs = self.term()?;
            expr = Expr {
                kind: Binary(BinaryExpr {
                    lhs: Box::new(expr),
                    rhs: Box::new(rhs),
                    operator: token.try_into()?,
                }),
            }
        }
        Ok(expr)
    }

    fn term(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let mut expr = self.factor()?;
        loop {
            let token = match self.peek_type()? {
                TokenType::Minus | TokenType::Plus => self.advance()?,
                _ => break,
            };
            let rhs = self.factor()?;
            expr = Expr {
                kind: Binary(BinaryExpr {
                    lhs: Box::new(expr),
                    rhs: Box::new(rhs),
                    operator: token.try_into()?,
                }),
            }
        }
        Ok(expr)
    }

    fn factor(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let mut expr = self.unary()?;
        loop {
            let token = match self.peek_type()? {
                TokenType::Slash | TokenType::Star => self.advance()?,
                _ => break,
            };
            let rhs = self.unary()?;
            expr = Expr {
                kind: Binary(BinaryExpr {
                    lhs: Box::new(expr),
                    rhs: Box::new(rhs),
                    operator: token.try_into()?,
                }),
            }
        }
        Ok(expr)
    }

    fn unary(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let expr = match self.peek_type()? {
            TokenType::Bang | TokenType::Minus => {
                let operator = self.advance()?.try_into()?;
                let expr = Box::new(self.unary()?);
                Expr {
                    kind: ExprKind::Unary(UnaryExpr { operator, expr }),
                }
            }
            _ => self.primary()?,
        };
        Ok(expr)
    }

    fn primary(&mut self) -> Result<Expr<'src>, SyntaxError> {
        let expr = match self.peek_type()? {
            TokenType::Keyword(Keyword::True) => Expr {
                kind: ExprKind::Literal(Boolean(true)),
            },
            TokenType::Keyword(Keyword::False) => Expr {
                kind: ExprKind::Literal(Boolean(false)),
            },
            TokenType::Keyword(Keyword::Nil) => Expr {
                kind: ExprKind::Literal(Nil),
            },
            TokenType::Literal(_) => {
                let literal: ast::Literal = self.advance()?.try_into()?;
                Expr {
                    kind: ExprKind::Literal(literal),
                }
            }
            TokenType::LeftParen => {
                self.consume(TokenType::LeftParen)?;
                let expr = self.expression()?;
                self.consume(TokenType::RightParen)?;
                Expr {
                    kind: ExprKind::Grouped(Box::new(expr)),
                }
            }
            _ => return Err(SyntaxError::Expect("expression")),
        };
        Ok(expr)
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use crate::src;
    const SNAPSHOT_OUTPUT_BASE: &'static str = "../snapshots/parser/snapshots-outputs";
    const SNAPSHOT_INPUT_BASE: &'static str = "../snapshots/parser/snapshots-inputs";

    #[test]
    fn primary() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "primary.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).primary()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn unary() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "unary.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).unary()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn factor() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "factor.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).factor()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn term() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "term.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).term()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn comparison() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "comparison.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).comparison()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn equality() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "equality.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).equality()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn expression() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "expression.lox");
            let asts = src.split('\n').map(|line| {
                Parser::new(line).expression()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }

    #[test]
    fn statements() {
        insta::with_settings!({snapshot_path => SNAPSHOT_OUTPUT_BASE},{
            let src = src!(SNAPSHOT_INPUT_BASE, "statements.lox");
            let asts = src.split('\n').map(|line| {
                println!("stmt => {}", &line);
                Parser::new(line).statement()
            }).collect::<Vec<_>>();
            insta::assert_debug_snapshot!(asts);
        })
    }
}
