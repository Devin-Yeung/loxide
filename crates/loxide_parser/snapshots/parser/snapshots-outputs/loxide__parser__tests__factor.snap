---
source: src/parser.rs
expression: asts
---
[
    Ok(
        Expr {
            kind: Binary(
                BinaryExpr {
                    lhs: Expr {
                        kind: Binary(
                            BinaryExpr {
                                lhs: Expr {
                                    kind: Literal(
                                        Number(
                                            2.0,
                                        ),
                                    ),
                                },
                                rhs: Expr {
                                    kind: Literal(
                                        Number(
                                            3.0,
                                        ),
                                    ),
                                },
                                operator: Star,
                            },
                        ),
                    },
                    rhs: Expr {
                        kind: Literal(
                            Number(
                                6.0,
                            ),
                        ),
                    },
                    operator: Slash,
                },
            ),
        },
    ),
]
